docs/proposals-accepted/20231114-split-store-gateway.md

```diff
@@ -0,0 +1,218 @@
+  repeated Label labels = 2;
+  repeated ChunkMeta chunks = 3;
+}
+
+message ChunkMeta {
+  int64 mint = 1;
+  int64 maxt = 2;
+  uint64 ref = 3;
+}
+```
+
+### `ChunkStore`
+
+The `ChunkStore` will implement a bi-directional streaming API. The Chunks API accepts a stream of chunks request containing blockID and the chunkRefs. After receiving all the requests the chunk store will return the response for each request in the same order.
+
+#### API
+
+#### Service
+
+```
+service ChunkStore {
+    rpc Chunks(stream ChunksRequest) returns (stream ChunksResponse);
+}
+```
+
+#### Models
+
+```
+message ChunksRequest {
+    string blockId = 1;
+    repeated uint64 chunkref = 2;
+}
+
+message ChunksResponse {
+    repeated typespb.AggrChunk chunks = 1;
+}
+```
+
+### Querier
+
+#### Fetching series and chunks
+
+The querier will first make a `Select()` call to the `IndexStore` to retrieve the Series labels and the metadata about the chunks. The querier would make concurrent calls to multiple `ChunkStore`s to read the corresponding chunks. A `SeriesSet` will be created by stitching together the chunks with the corresponding series.
+
+#### Adapting the Select()/Chunks() into Series()
+
+Querier fetches data from Stores using Store API, which uses Series API to fetch series labels and chunks at the same time. The new Select and Chunks API are not compatible with Series API and we don’t want to introduce breaking changes.
+
+To allow querying `IndexStore` and `ChunkStore` using `Select()` and `Chunks()` requests, we can add an adapter in querier which exposes Store API, but uses `Select()` and `Chunks()` API to fetch data instead.
+
+The adapter can be implemented as a `storepb.StoreClient`. https://github.com/thanos-io/thanos/blob/v0.32.5/pkg/store/proxy.go#L49
+
+#### Merging series from multiple blocks
+
+Currently the `Series()` API implemented by `Store` allows querying multiple blocks at the same time. The proxy heap is used to merge the series read from multiple blocks and remove duplicate series and identical chunks. This requires some rethought and a few possibilities are described below.
+
+#### Merge both series and chunks on querier
+
+The `IndexStore` only supports querying a single block per request. Each Thanos querier would make separate `Select()` request to the `IndexStore` for each block. The merging and the de-dupes will be done on the querier after getting the chunks.
+
+#### Merge series on IndexStore and de-dupe chunks on querier [preferred]
+
+The `IndexStore` will allow querying multiple blocks per request. Only the series dedupes would be performed on the `IndexStore`. Identical chunks would be removed on the querier after chunks are fetched from the `ChunkStore`
+
+### Endpoint Discovery
+
+`Info` API will be exposed in `IndexStore` so that Querier can discover index store the same way as discovering current stores. `ChunkStore` is stateless and it only exposes `Chunks()` gRPC API. Querier can use DNS service discovery for it.
```

Comment:
Maybe it would be cool to have a separate table in the UI with the known chunk stores?